basesandbox/src/ipc.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pub mod domain_socket;
+pub mod intra;
+
+pub trait IpcSend: Send {
+    /// It might block until counterparty's recv(). Even if not, the order is still guaranteed.
+    fn send(&self, data: &[u8]);
+}
+
+#[derive(Debug, PartialEq)]
+pub enum RecvError {
+    TimeOut,
+    Termination,
+}
+
+pub trait Terminate: Send {
+    /// Wake up block on recv with a special flag
+    fn terminate(&self);
+}
+
+pub trait IpcRecv: Send {
+    type Terminator: Terminate;
+
+    /// Returns Err only for the timeout or termination wake-up(otherwise panic)
+    fn recv(&self, timeout: Option<std::time::Duration>) -> Result<Vec<u8>, RecvError>;
+    /// Create a terminate switch that can be sent to another thread
+    fn create_terminate(&self) -> Self::Terminator;
+}
+
+/// One possible abstraction for a server-client style of `Linker`
+/// TwoWay means you there are exactly two communcations to establish IPC.
+/// And moreover, those two communications have no dependencies, so it trivially returns
+/// two arguments (to be sent).
+pub trait TwoWayInitialize {
```

Comment:
I'm still not sure if `TwoWayInitialize` is a necessary abstraction here. It just produces a pair of messages and there is no direct relationship with `Ipc`.

Reply:
Could you please take a look of this https://github.com/CodeChain-io/foundry/pull/305#discussion_r409568704

Reply:
If `Linker` is used that way, I still don't get why `TwoWayInitialize` should be separate. Do you have any strong requirement to reuse single impl of `TwoWayInitialize` for multiple different `Ipc`s?

Reply:
@dynaxis I don't get what you mean. `TwoWayInitialize` is implemented at least once for every IPCs.

Reply:
1. `TwoWayInitialize` just generates a pair of arguments. So see no reason why it should be a separate `object` at all.
1. `TwoWayInitialize` is related to an IPC mechanism. And therefore implemented for **each IPC**. Then is there any reason to not merge it into `Ipc`? It's anyway specific to an IPC mechanism. Note that you can have trait level functions without taking a `self`.
1. If it's ever important to have a linker separate (in your design I mean. The linker in my API design is a bit different), I would rather have made the linking process more into the linker concept, rather than keeping a separate function such as `execute`. Then you can hide more from the outside.

The above are all of my points.

Reply:
@dynaxis At first in the PR, the `Linker` is assumed to be object so that it can stay alive after linkage, for cleaning up purpose. (current generic function of `execute` can take advantage of possible `linker`'s `Drop`) And there was actually such `Drop` impl for `DomainSocket`, but it was turned out to be ok to be just in each end, as @majecty requested. So yes, you're right on this point. For current two implementations of `Ipc`, it is ok with `TwoWayInitialize` compensated with a single static (no `&self`) function. If this bothers you, I'll just apply your suggestion.